sycl/doc/extensions/AsyncGroupCopy/async_group_copy.asciidoc

```diff
@@ -0,0 +1,115 @@
+# Async Group Copy Extension for DPC++: SYCL_EXT_ONEAPI_ASYNC_GROUP_COPY
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+:dpcpp: pass:[DPC++]
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+Copyright (c) 2021-2021 Intel Corporation.  All rights reserved.
+
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension is written against the SYCL 2020 revision 3 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+## Contributors
+
+* Gordon Brown
+* Tadej Ciglaric
+* Jack Kirk
+* Finlay Marno
+
+## Introduction
+
+This document proposes that `async_work_group_copy` be deprecated and replaced 
+with `async_group_copy`. `async_group_copy` is a non-member function, in line
+with the other group functions, which generalizes `async_work_group_copy` to
+also work with sub-groups.
```

Comment:
Bikeshedding: It didn't occur to me until today that the name `group_copy` is aligned with the "Group functions", and it would probably be called `joint_group_copy` if it were defined as a "Group algorithm".

Is the intent here for this to be a function or an algorithm?

Reply:
I'm actually a little bit confused about where `async_group_copy` fits in. It definitely requires all the work-items to call the function with the same parameters, but I don't see a necessity for the work-items to do the copies synchronously. 

We are a bit unsure on the wait as well. We don't think it should necessary synchronize the group, so maybe doesn't actually need the group as an argument. The memory a single work-item copied should be consistent for that work item after wait, but there doesn't need to be guarantees about the memory that other work-items work on. I guess the question then is "what memory did the work-item copy?".

Reply:
I just double-checked the OpenCL definitions, and they require both the copy and wait call to be encountered by all work-items in the group: https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_C.html#async-copies.

Given what you said earlier about it being a low-level functionality, I think positioning the copy and wait functions as "Group functions" makes the most sense.  Viewing these as a collective operation allows for more implementations (e.g. an implementation can sit atop OpenCL), and avoids the complexity you mentioned -- as soon as you specify that certain work-items copy specific memory locations, you might prevent this operation from being implemented effectively on some hardware.

The semantics that make the most sense to me are:
- `async_group_copy` must be encountered by all work-items in converged control flow and acts as a synchronization point with an acquire fence.  (Otherwise, different work-items might disagree on what values they're copying).
- `wait` must be encountered by all work-items in converged control flow and acts as a synchronization point with a release fence. (Otherwise, different work-items might disagree on what values were copied).

The OpenCL specification is clear that these operations in OpenCL _don't_ imply any sort of fencing, but that seems poorly defined to me.

Reply:
Ok, I agree on copy and wait being collective operations. As you say, I don't want to specify exactly what data each work-items copies.

Since we can't tell what work each work-item did, I think adding a synchronization point with an acquire-release fence to `wait` makes sense. Otherwise you would just always need one for correctness.

I am a little reluctant to add the synchronization point with an acquire-release fence to `async_group_copy` since I can see that being used as either the first or last function call in a kernel and it would be shame to have a wasted sync.

Reply:
> I am a little reluctant to add the synchronization point with an acquire-release fence to async_group_copy since I can see that being used as either the first or last function call in a kernel and it would be shame to have a wasted sync.

I see where you're coming from and I share the reluctance.  It feels like the work-items should be able to start opportunistically copying the part of the range assigned to them as soon as they hit the `async_group_copy` call without a sync, but I don't think that works in practice.  I think in real-life use-cases you'd need to add an explicit sync before the group copy (just like with the wait).

Consider this example:
```c++
// Each work-item writes a few elements to local memory with a stride
local[it.get_local_id()] = x;
local[it.get_local_id() + it.get_local_range()] = y;

// Running this without a sync requires this work-item to be responsible for
// copying the same values as written above
async_group_copy(g, local, global, 2 * it.get_local_range());
```

In general I don't see how you can provide the necessary guarantee without specifying a relationship between work-items and the values that they copy.  Saying that the copy is implicitly strided across work-items would address the example above, but would then break in a case like this:

```c++
// Each work-item writes a few contiguous elements to local memory
local[it.get_local_id() * it.get_local_range() + 0] = x;
local[it.get_local_id() * it.get_local_range() + 1] = y;

// Running this without a sync requires this work-item to be responsible for
// copying the same values as written above
async_group_copy(g, local, global, 2 * it.get_local_range());
```

Reply:
I agree this is a potential pitfall but the programmer should know that work-items will have a potentially inconsistent view of memory after they have been writing to it. I don't believe `joint_reduce` synchronizes the work-items first, the programmer is expected to add in a barrier if necessary.

Reply:
> I agree this is a potential pitfall but the programmer should know that work-items will have a potentially inconsistent view of memory after they have been writing to it.

This being the case, is there any way a programmer can use `async_group_copy` _without_ synchronizing the work-items first?  They must always write to the local memory before copying it to global memory (otherwise its values are undefined), and the only way to ensure that the work-items have a consistent view of memory prior to calling `async_group_copy` would be to synchronize them.  If the only safe way to use this function is to precede it with a barrier, why not put the barrier in the function itself?

> I don't believe joint_reduce synchronizes the work-items first, the programmer is expected to add in a barrier if necessary.

By my reading of the specification, the group algorithms do synchronize the work-items first, so this part of the specification may require clarification.

[Section 4.17.4. Group algorithms library](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:algorithms) says:
> All algorithms act as group functions (as defined in Section 4.17.3), inheriting all restrictions of group functions.

[Section 4.17.3. Group functions](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:group-functions) then says:
> These group functions act as synchronization points and must be encountered in converged control flow by all work-items in the group.

The meaning of "synchronization points" here isn't expanded upon, but the standard C++ [_synchronizes-with_](http://eel.is/c++draft/intro.races#def:synchronize_with) relationship implies some agreement about the state of memory.  The definition of group barrier in the glossary says:
> A synchronization function within a group of work items. All the work items of a group must execute the barrier construct before any work-item continues execution beyond the barrier. Additionally all work-items in the group execute a release mem-fence prior to synchronizing at the barrier, all work-items in the group execute an acquire mem-fence after synchronizing at the barrier, and there is an implicit synchronization between these acquire and release fences as if through an atomic operation on an atomic object internal to the barrier implementation.

...so I think it's possible to read this as saying that there are appropriate fences and barriers at the beginning and end of each group algorithm.

Reply:
The case would be at the beginning of a kernel, when loading memory from global to local. For CUDA, there is actually no asynchronous load from local to global!

From https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async :

> It is important to note that while this is a memcpy in the general case, it is only asynchronous if the source is global memory and the destination is shared memory and both can be addressed with 16, 8, or 4 byte alignments.

Maybe it is the case that `async_group_copy` is neither a group function or a group algorithm due to the restrictions on them.

I don't want to distract but looking at the implementation for `joint_reduce`, I don't see any synchronization before the `sycl::detail::for_each` is done https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/group_algorithm.hpp#L231.

Reply:
> I don't want to distract but looking at the implementation for `joint_reduce`, I don't see any synchronization before the `sycl::detail::for_each` is done https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/group_algorithm.hpp#L231.

Hm, I didn't realize that, thanks for pointing it out.

I think we need the specification to be clarified before we can actually decide what the right thing to do here is.  I've opened an issue at Khronos: https://gitlab.khronos.org/sycl/Specification/-/issues/576.

Reply:
> The case would be at the beginning of a kernel, when loading memory from global to local. For CUDA, there is actually no asynchronous load from local to global!
> 
> From [docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async) :
> 
> > It is important to note that while this is a memcpy in the general case, it is only asynchronous if the source is global memory and the destination is shared memory and both can be addressed with 16, 8, or 4 byte alignments.

I guess it is somehow an optimization and implementation detail.
For example Xilinx OpenCL implementation implements the asynchronous copy as just a synchronous copy.
This is also correct. :-)
So we might also expose things in SYCL even if the CUDA back-end cannot take advantage from it.

Reply:
> > The case would be at the beginning of a kernel, when loading memory from global to local. For CUDA, there is actually no asynchronous load from local to global!
> > From [docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async) :
> > > It is important to note that while this is a memcpy in the general case, it is only asynchronous if the source is global memory and the destination is shared memory and both can be addressed with 16, 8, or 4 byte alignments.
> 
> I guess it is somehow an optimization and implementation detail. For example Xilinx OpenCL implementation implements the asynchronous copy as just a synchronous copy. This is also correct. :-) So we might also expose things in SYCL even if the CUDA back-end cannot take advantage from it.

cc @Pennycook @keryell 

The initial accompanying implementation to this proposal did make use of the asynchronous copy using the Direct Memory Access hardware for copying from global to shared memory introduced in sm_80 Ampere architecture, but it did not fully allow the overlapping of these DMA memory copies with computes, since the wait_for implementation waited for all outstanding asynchronous copies.  We now plan to introduce functionality similar to ["wait_prior"](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-wait) from cuda runtime so that we can get more "free" computes during the async copies, by executing multiple async copies and only waiting for those necessary for a following compute.

This poses some obvious questions/problems because the cuda backend does not return anything that can be translated into a device_event from async_group_copy (as discussed [here](https://github.com/intel/llvm/pull/4903#discussion_r743803534) the backend instead takes an argument N telling it how many async copies to *NOT* wait for).  Currently there is no backend in DPC++ (or any other SYCL implementation as far as I am aware!) that actually uses async_work_group_copy asynchronously, and I am not sure that any GPU backend beyond Nvidia actually has hardware to support it! It looks like it is mostly CPU and FPGA backends that might have supported asynchronous async_work_group_copy via DMA, although this is just a guess from some google searches, so if anyone can clarify this I'd appreciate it! 
The OpenCL async_work_group_copy is more than a decade old so it would be good to confirm that it is still being actively used, since I am a little uncomfortable that it is incompatible with the only backend in DPC++ currently supporting asynchronous copies to shared memory (cuda).  If in the future other backends plan to implement asynchronous copies from global to local memory, it would be an opportune moment to consider whether they will make use of the existing async_work_group_copy/device_event interface/workflow.

Assuming that we do keep the existing device_event workflow we need to work out how we can make a compatible cuda implementation.
One option would be to introduce a new optional template parameter to sycl::wait_for that can be used by (at least) the cuda backend to *NOT* wait for numEvents, and only wait for the other totalEvents - numEvents asynchronous copies, something like:
```
template <int numEvents, typename Group, typename... eventT>
std::enable_if_t<(sycl::is_group_v<Group>) &&
                 (std::is_same_v<eventT, async_copy_event<Group>> && ...)>
wait_for(Group, eventT... Events) {
  #ifdef __NVPTX__
  if constexpr (numEvents == 0)
  {
    (__spirv_GroupWaitEvents(detail::group_to_scope<Group>(), 0, &Events.Event),
   ...);
  }
  else{
  __nvvm_cp_async_wait_group(numEvents);
    __spirv_ControlBarrier(sycl::detail::group_barrier_scope<Group>::Scope,
                         0,
                         __spv::MemorySemanticsMask::AcquireRelease |
                             __spv::MemorySemanticsMask::SubgroupMemory |
                             __spv::MemorySemanticsMask::WorkgroupMemory |
                             __spv::MemorySemanticsMask::CrossWorkgroupMemory);
  }
   #else
   (__spirv_GroupWaitEvents(detail::group_to_scope<Group>(), 0, &Events.Event),
   ...);
   #endif
}
```

If this is not acceptable I think we will have to somehow add a list of the order of the async copy device_event events to the cuda backend so that we can map each device_event to the correct N numEvents. 

I am assuming that the async_work_group_copy from the SYCL spec is still useful for some backends in DPC++ to achieve true asynchronous copies.  I would be more comfortable actually knowing this.  If on the other hand async_work_group_copy is not appropriate then we should have a conversation about what a more appropriate interface would be for general DPC++ backends.

Reply:
> > The case would be at the beginning of a kernel, when loading memory from global to local. For CUDA, there is actually no asynchronous load from local to global!
> > From [docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-memcpy-async) :
> > > It is important to note that while this is a memcpy in the general case, it is only asynchronous if the source is global memory and the destination is shared memory and both can be addressed with 16, 8, or 4 byte alignments.
> 
> I guess it is somehow an optimization and implementation detail. For example Xilinx OpenCL implementation implements the asynchronous copy as just a synchronous copy. This is also correct. :-) So we might also expose things in SYCL even if the CUDA back-end cannot take advantage from it.

cc @Pennycook @keryell

The initial accompanying implementation to this proposal did make use of the asynchronous copy using the Direct Memory Access hardware for copying from global to shared memory introduced in sm_80 Ampere architecture, but it did not fully allow the overlapping of these DMA memory copies with computes, since the wait_for implementation waited for all outstanding asynchronous copies. We now plan to introduce functionality similar to ["wait_prior"](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#collectives-cg-wait) from cuda runtime so that we can get more "free" computes during the async copies, by executing multiple async copies and only waiting for those necessary for a following compute.

This poses some obvious questions/problems because the cuda backend does not return anything that can be translated into a device_event from async_group_copy (as discussed https://github.com/intel/llvm/pull/4903#discussion_r743803534 the backend instead takes an argument N telling it how many async copies to NOT wait for). Currently there is no backend in DPC++ (or any other SYCL implementation as far as I am aware!) that actually uses async_work_group_copy asynchronously, and I am not sure that any GPU backend beyond Nvidia actually has hardware to support it! It looks like it is mostly CPU and FPGA backends that might have supported asynchronous async_work_group_copy via DMA, although this is just a guess from some google searches, so if anyone can clarify this I'd appreciate it!
The OpenCL async_work_group_copy is more than a decade old so it would be good to confirm that it is still being actively used, since I am a little uncomfortable that it is incompatible with the only backend in DPC++ currently supporting asynchronous copies to shared memory (cuda). If in the future other backends plan to implement asynchronous copies from global to local memory, it would be an opportune moment to consider whether they will make use of the existing async_work_group_copy/device_event interface/workflow.

Assuming that we do keep the existing device_event workflow we need to work out how we can make a compatible cuda implementation.
One option would be to introduce a new optional template parameter to sycl::wait_for that can be used by (at least) the cuda backend to NOT wait for numEvents, and only wait for the other totalEvents - numEvents asynchronous copies, something like:

template <int numEvents, typename Group, typename... eventT>
std::enable_if_t<(sycl::is_group_v<Group>) &&
                 (std::is_same_v<eventT, async_copy_event<Group>> && ...)>
wait_for(Group, eventT... Events) {
  #ifdef __NVPTX__
  if constexpr (numEvents == 0)
  {
    (__spirv_GroupWaitEvents(detail::group_to_scope<Group>(), 0, &Events.Event),
   ...);
  }
  else{
  __nvvm_cp_async_wait_group(numEvents);
    __spirv_ControlBarrier(sycl::detail::group_barrier_scope<Group>::Scope,
                         0,
                         __spv::MemorySemanticsMask::AcquireRelease |
                             __spv::MemorySemanticsMask::SubgroupMemory |
                             __spv::MemorySemanticsMask::WorkgroupMemory |
                             __spv::MemorySemanticsMask::CrossWorkgroupMemory);
  }
   #else
   (__spirv_GroupWaitEvents(detail::group_to_scope<Group>(), 0, &Events.Event),
   ...);
   #endif
}

If this is not acceptable I think we will have to somehow add a list of the order of the async copy device_event events to the cuda backend so that we can map each device_event to the correct N numEvents.

I am assuming that the async_work_group_copy from the SYCL spec is still useful for some backends in DPC++ to achieve true asynchronous copies. I would be more comfortable actually knowing this. If on the other hand async_work_group_copy is not appropriate then we should have a conversation about what a more appropriate interface would be for general DPC++ backends.

Reply:
> > I agree this is a potential pitfall but the programmer should know that work-items will have a potentially inconsistent view of memory after they have been writing to it.
> 
> This being the case, is there any way a programmer can use `async_group_copy` _without_ synchronizing the work-items first? They must always write to the local memory before copying it to global memory (otherwise its values are undefined), and the only way to ensure that the work-items have a consistent view of memory prior to calling `async_group_copy` would be to synchronize them. If the only safe way to use this function is to precede it with a barrier, why not put the barrier in the function itself?
> 
> > I don't believe joint_reduce synchronizes the work-items first, the programmer is expected to add in a barrier if necessary.
> 
> By my reading of the specification, the group algorithms do synchronize the work-items first, so this part of the specification may require clarification.
> 
> [Section 4.17.4. Group algorithms library](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:algorithms) says:
> 
> > All algorithms act as group functions (as defined in Section 4.17.3), inheriting all restrictions of group functions.
> 
> [Section 4.17.3. Group functions](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:group-functions) then says:
> 
> > These group functions act as synchronization points and must be encountered in converged control flow by all work-items in the group.
> 
> The meaning of "synchronization points" here isn't expanded upon, but the standard C++ [_synchronizes-with_](http://eel.is/c++draft/intro.races#def:synchronize_with) relationship implies some agreement about the state of memory. The definition of group barrier in the glossary says:
> 
> > A synchronization function within a group of work items. All the work items of a group must execute the barrier construct before any work-item continues execution beyond the barrier. Additionally all work-items in the group execute a release mem-fence prior to synchronizing at the barrier, all work-items in the group execute an acquire mem-fence after synchronizing at the barrier, and there is an implicit synchronization between these acquire and release fences as if through an atomic operation on an atomic object internal to the barrier implementation.
> 
> ...so I think it's possible to read this as saying that there are appropriate fences and barriers at the beginning and end of each group algorithm.

I think that we agree that https://gitlab.khronos.org/sycl/Specification/-/issues/576 should clarify the meaning of 

"These group functions act as synchronization points and must be encountered in converged control flow by all work-items in the group."

As Finlay points out, the current implementation of the reduce_over_group collective is not consistent with an interpretation of the above statement as meaning that group collectives require fencing by definition.

As you point out the OpenCL definition of `async_work_group_copy` doesn't imply any sort of fencing.  I believe that this proposal defines `async_group_copy` as a straightforward generalization of `async_work_group_copy` to support sub-groups.  The behaviour of `async_group_copy` (or whatever its final name is chosen to be!) in the `group` case should match the behaviour of async_work_group_copy.  This should mean that if `async_group_copy` needs to be altered then `async_work_group_copy` should be identically altered (They would have to be atm, since they use the same SPIR-V implementation!).

There are strong performance reasons for not synchronizing with `sycl::barrier` prior to a call to to `async_group_copy`.  I find that if I add an additional call to `barrier` prior to `group_async_copy` when copying 8KB of data from global to shared using the max work group size, the sm_80 DMA performance is worse than if I did a non-asynchronous copy on the same device, assuming that there is no useful compute that can be made whilst the copy is being performed by the DMA without using the new data that is being copied, even though the DMA route copies from global to shared more efficiently by avoiding the usage of registers.
It is worth noting that although the CUDA implementation of the asynchronous copy does not require the work-items to be converged, there is a considerable performance advantage to the work-items being converged: see this discussion on `warp entanglement`: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-entanglement-commit.
Our existing ptx implementation is consistent with the CUDA runtime where it is also left up to the user to ensure that the threads are converged.
Taking this and the OpenCL definition of `work_group_async_copy` into account, given that I am not aware of any DMA setups other than those supported by OpenCL and CUDA, it seems likely that for the purposes of `group_async_copy` the barrier prior to `group_async_copy` will not be necessary unless the user has explicitly diverged the work-items.  However since `group_async_copy` may fall under the banner of sycl `group functions` in the future, if it is decided to be important for other sycl `group functions`, we may have to add a barrier to the beginning of `group_async_copy`.  However I don't think that this in itself can be a blocker to merging this proposal, particularly since it is aimed for the `oneapi::experimental` namespace.

As an aside for the bigger picture it is worth mentioning that asynchronous copies are most advantageous when copying as much memory as possible in a single kernel to shared memory, and when multiple memory copies (often roughly of size 8KB) can be initialised in tandem, with concurrent computes (and therefore waits) on subsets of these data copies while other subsets are left in flight (I think this may be referred to as "double buffering").  The most obvious application for this in DPC++ applications that comes to mind for me is for Matrix MMA kernels, however the current interface of work_group_async_copy/group_async_copy is incompatible for such an application, since it would be necessary to specify precisely the data that each work-item will copy (I don't see that this could not be done in a sub-group/group convergent way).  Memory copies to shared memory can be a bottleneck for such kernels.